File: src/main/java/com/starshootercity/OriginsAddon.java
package com.starshootercity;

import com.starshootercity.abilities.Ability;
import com.starshootercity.abilities.AbilityRegister;
import net.kyori.adventure.resource.ResourcePackInfo;
import org.bukkit.plugin.java.JavaPlugin;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.util.List;

public abstract class OriginsAddon extends JavaPlugin {
    private static OriginsAddon instance;

    public static OriginsAddon getInstance() {
        return instance;
    }

    @Override
    public final void onEnable() {
        instance = this;
        onRegister();
        AddonLoader.register(this);
        for (Ability ability : getAbilities()) {
            AbilityRegister.registerAbility(ability, this);
        }
        PackApplier.addResourcePack(this, getResourcePackInfo());
    }

    public @Nullable ResourcePackInfo getResourcePackInfo() {
        return null;
    }

    @Override
    public @NotNull File getFile() {
        return super.getFile();
    }

    public void onRegister() {}

    public abstract @NotNull String getNamespace();

    public @NotNull List<Ability> getAbilities() {
        return List.of();
    }
}

File: core/src/main/kotlin/ru/turbovadim/OriginsAddon.kt
package ru.turbovadim

import ru.turbovadim.abilities.Ability
import ru.turbovadim.abilities.AbilityRegister
import ru.turbovadim.abilities.custom.ToggleableAbility
import ru.turbovadim.events.PlayerSwapOriginEvent
import ru.turbovadim.packetsenders.OriginsRebornResourcePackInfo
import net.kyori.adventure.key.Key
import org.bukkit.entity.Player
import org.bukkit.plugin.java.JavaPlugin
import java.io.File

abstract class OriginsAddon : JavaPlugin() {

    companion object {
        private lateinit var instance: OriginsAddon
        fun getInstance(): OriginsAddon = instance
    }

    open fun shouldOpenSwapMenu(): SwapStateGetter? = null

    open fun shouldAllowOriginSwapCommand(): SwapStateGetter? = null

    open fun getAbilityOverride(): KeyStateGetter? = null

    interface SwapStateGetter {
        fun get(player: Player, reason: PlayerSwapOriginEvent.SwapReason): State
    }

    interface KeyStateGetter {
        fun get(player: Player, key: Key): State
    }

    @Suppress("unused")
    enum class State {
        ALLOW,
        DEFAULT,
        DENY
    }

    final override fun onEnable() {
        instance = this
        onRegister()
        AddonLoader.register(this)
        for (ability in getAbilities()) {
            if (ability is ToggleableAbility && !ability.shouldRegister()) continue
            AbilityRegister.registerAbility(ability, this)
        }
        getResourcePackInfo()?.let { PackApplier.addResourcePack(this, it) }
        afterRegister()
    }

    open fun getResourcePackInfo(): OriginsRebornResourcePackInfo? = null

    public override fun getFile(): File = super.getFile()

    open fun onRegister() {}

    open fun afterRegister() {}

    abstract fun getNamespace(): String

    open fun getAbilities(): List<Ability> = listOf()
}
